File: PointPilot/Models/Metrics.swift
import Foundation

/// One past recommendation, kept so the dashboard can show totals.
///
/// These are demo aggregates: the app has no account, no bank connection and no
/// server, so the history is a fixed sample set rather than a record of the
/// user's real spending. It is labelled as demo data wherever it is shown.
struct HistoryEntry: Identifiable, Hashable, Sendable {
    let id: String
    let merchantName: String
    let category: RewardCategory
    /// What was spent.
    let amount: Decimal
    /// The card the user actually used.
    let usedCardID: String
    let usedCardName: String
    /// Value actually earned on this purchase, as the engine computes it.
    let earned: Decimal
    /// The best value available in the wallet. Equal to `earned` when the user
    /// took the recommendation; greater when a weaker card was used.
    let bestPossible: Decimal
    /// How long ago the purchase happened, in days.
    let daysAgo: Int

    /// Value left behind by not using the best card.
    var missedValue: Decimal { max(0, bestPossible - earned) }

    /// True when a better card existed and was not used.
    var usedSuboptimalCard: Bool { missedValue > 0 }
}

/// Aggregated dashboard figures derived from a list of history entries.
///
/// Pure value type with no I/O, so the arithmetic is testable on fixed fixtures
/// and the view stays declarative.
struct MetricsSummary: Hashable, Sendable {
    struct CardStanding: Identifiable, Hashable, Sendable {
        let cardID: String
        let cardName: String
        let cardAccentHex: String
        let earned: Decimal
        let uses: Int

        var id: String { cardID }
    }

    let entries: [HistoryEntry]

    init(entries: [HistoryEntry]) {
        // Newest first so "this month" and the activity list agree on ordering.
        self.entries = entries.sorted { $0.daysAgo < $1.daysAgo }
    }

    /// Entries inside the trailing 30-day window the dashboard calls "this month".
    var recentEntries: [HistoryEntry] { entries.filter { $0.daysAgo <= 30 } }

    var totalEarned: Decimal {
        recentEntries.reduce(Decimal.zero) { $0 + $1.earned }
    }

    var totalSpent: Decimal {
        recentEntries.reduce(Decimal.zero) { $0 + $1.amount }
    }

    /// The "money left on the table" figure: value given up by using a card
    /// other than the one the engine would have recommended.
    var missedValue: Decimal {
        recentEntries.reduce(Decimal.zero) { $0 + $1.missedValue }
    }

    /// Effective return across all tracked spend, as a percentage (1.4 == 1.4%).
    var effectiveRate: Decimal {
        guard totalSpent > 0 else { return 0 }
        return Currency.rounded(totalEarned / totalSpent * 100)
    }

    /// Cards ranked by value earned, highest first. Ties break on use count,
    /// then name, so the order is stable across renders.
    var standings: [CardStanding] {
        var earnedByCard: [String: Decimal] = [:]
        var usesByCard: [String: Int] = [:]
        var nameByCard: [String: String] = [:]

        for entry in recentEntries {
            earnedByCard[entry.usedCardID, default: 0] += entry.earned
            usesByCard[entry.usedCardID, default: 0] += 1
            nameByCard[entry.usedCardID] = entry.usedCardName
        }

        return earnedByCard
            .map { cardID, earned in
                CardStanding(
                    cardID: cardID,
                    cardName: nameByCard[cardID] ?? cardID,
                    cardAccentHex: Self.accentHex(forCardID: cardID),
                    earned: earned,
                    uses: usesByCard[cardID] ?? 0
                )
            }
            .sorted { lhs, rhs in
                if lhs.earned != rhs.earned { return lhs.earned > rhs.earned }
                if lhs.uses != rhs.uses { return lhs.uses > rhs.uses }
                return lhs.cardName < rhs.cardName
            }
    }

    /// The card earning the most across the window, or nil when there is no history.
    var topCard: CardStanding? { standings.first }

    /// Spend grouped by category, largest first — drives the breakdown bars.
    var categoryBreakdown: [(category: RewardCategory, amount: Decimal)] {
        var totals: [RewardCategory: Decimal] = [:]
        for entry in recentEntries {
            totals[entry.category, default: 0] += entry.amount
        }
        return totals
            .map { (category: $0.key, amount: $0.value) }
            .sorted { lhs, rhs in
                if lhs.amount != rhs.amount { return lhs.amount > rhs.amount }
                return lhs.category.rawValue < rhs.category.rawValue
            }
    }

    /// Accent color for a card, so the dashboard can render it consistently
    /// with the wallet without depending on the repository.
    private static func accentHex(forCardID cardID: String) -> String {
        switch cardID {
        case SampleDataRepository.amexGoldID: return "#B99A5B"
        case SampleDataRepository.sapphirePreferredID: return "#2F6BA8"
        case SampleDataRepository.ventureID: return "#8C2F39"
        default: return "#6E7F96"
        }
    }
}

File: web/src/lib/metrics.ts
import { roundToCents, ratio, type Money } from './money'
import { evaluate, recommend } from './engine'
import { sampleCards, sampleDiningProgram, sampleMerchants } from './sampleData'
import type { HistoryEntry, MetricsSummary, RewardCategory } from './types'

/**
 * The dashboard aggregates — the web port of `Metrics.swift`.
 *
 * The demo history is *replayed through the real engine* rather than
 * hand-written, so the totals on the dashboard can never contradict what the
 * finder computes for the same purchase. That is the same guarantee the iOS app
 * makes, and `parity.test.ts` pins both to the same figures.
 */

interface HistorySpec {
  id: string
  merchantQuery: string
  amount: Money
  usedCardID: string
  daysAgo: number
}

const MICRO = 1_000_000

const HISTORY_SPECS: HistorySpec[] = [
  { id: 'h-01', merchantQuery: 'Nobu', amount: 200 * MICRO, usedCardID: 'amex-gold', daysAgo: 2 },
  { id: 'h-02', merchantQuery: 'Whole Foods', amount: 140 * MICRO, usedCardID: 'chase-sapphire-preferred', daysAgo: 4 },
  { id: 'h-03', merchantQuery: 'Shake Shack', amount: 28 * MICRO, usedCardID: 'chase-sapphire-preferred', daysAgo: 6 },
  { id: 'h-04', merchantQuery: 'Shell', amount: 62 * MICRO, usedCardID: 'capital-one-venture', daysAgo: 9 },
  { id: 'h-05', merchantQuery: 'Chipotle', amount: 19 * MICRO, usedCardID: 'capital-one-venture', daysAgo: 12 },
  { id: 'h-06', merchantQuery: 'Apple Store', amount: 320 * MICRO, usedCardID: 'amex-gold', daysAgo: 15 },
  { id: 'h-07', merchantQuery: 'AMC Theatres', amount: 46 * MICRO, usedCardID: 'chase-sapphire-preferred', daysAgo: 21 },
  { id: 'h-08', merchantQuery: 'Starbucks', amount: 12 * MICRO, usedCardID: 'amex-gold', daysAgo: 26 },
]

/** Demo history, computed by the real engine. */
export function buildHistory(): HistoryEntry[] {
  const entries: HistoryEntry[] = []

  for (const spec of HISTORY_SPECS) {
    const result = recommend(spec.merchantQuery, spec.amount)
    const used = result.ranked.find((e) => e.card.id === spec.usedCardID)
    if (!used) continue

    entries.push({
      id: spec.id,
      merchantName: result.merchantName,
      category: result.category,
      amount: spec.amount,
      usedCardID: used.card.id,
      usedCardName: used.card.name,
      earned: used.totalValue,
      bestPossible: result.winner.totalValue,
      daysAgo: spec.daysAgo,
    })
  }

  return entries
}

const ACCENTS: Record<string, string> = {
  'amex-gold': '#B99A5B',
  'chase-sapphire-preferred': '#2F6BA8',
  'capital-one-venture': '#8C2F39',
}

/** Spend grouped by category, largest first — drives the breakdown bars. */
export function categoryBreakdown(
  entries: HistoryEntry[],
): { category: RewardCategory; amount: Money }[] {
  const totals = new Map<RewardCategory, Money>()
  for (const entry of entries) {
    totals.set(entry.category, (totals.get(entry.category) ?? 0) + entry.amount)
  }
  return [...totals.entries()]
    .map(([category, amount]) => ({ category, amount }))
    .sort((a, b) => {
      if (a.amount !== b.amount) return b.amount - a.amount
      return a.category < b.category ? -1 : 1
    })
}

/** Aggregates a list of history entries into the dashboard summary. */
export function summarize(entries: HistoryEntry[]): MetricsSummary {
  // Newest first so "this month" and the activity list agree on ordering.
  const sorted = [...entries].sort((a, b) => a.daysAgo - b.daysAgo)
  const recent = sorted.filter((e) => e.daysAgo <= 30)

  const totalEarned = recent.reduce((sum, e) => sum + e.earned, 0)
  const totalSpent = recent.reduce((sum, e) => sum + e.amount, 0)
  const missedValue = recent.reduce(
    (sum, e) => sum + Math.max(0, e.bestPossible - e.earned),
    0,
  )

  const earnedByCard = new Map<string, Money>()
  const usesByCard = new Map<string, number>()
  const nameByCard = new Map<string, string>()

  for (const entry of recent) {
    earnedByCard.set(entry.usedCardID, (earnedByCard.get(entry.usedCardID) ?? 0) + entry.earned)
    usesByCard.set(entry.usedCardID, (usesByCard.get(entry.usedCardID) ?? 0) + 1)
    nameByCard.set(entry.usedCardID, entry.usedCardName)
  }

  const standings = [...earnedByCard.entries()]
    .map(([cardID, earned]) => ({
      cardID,
      cardName: nameByCard.get(cardID) ?? cardID,
      cardAccentHex: ACCENTS[cardID] ?? '#6E7F96',
      earned,
      uses: usesByCard.get(cardID) ?? 0,
    }))
    .sort((a, b) => {
      if (a.earned !== b.earned) return b.earned - a.earned
      if (a.uses !== b.uses) return b.uses - a.uses
      return a.cardName < b.cardName ? -1 : 1
    })

  const effectiveRate =
    totalSpent > 0 ? roundToCents(ratio(totalEarned, totalSpent) * 100 * MICRO) : 0

  return {
    entries: sorted,
    recentEntries: recent,
    totalEarned,
    totalSpent,
    missedValue,
    effectiveRate,
    standings,
    categoryBreakdown: categoryBreakdown(recent),
  }
}

/** The app's demo dashboard summary. */
export function demoMetrics(): MetricsSummary {
  return summarize(buildHistory())
}

/**
 * A single card's reward for a purchase at a merchant, with no offer applied.
 *
 * Used where a screen wants to show one card's line without a full
 * recommendation — kept here so callers do not reach into engine internals.
 */
export function cardBaseReward(
  cardID: string,
  amount: Money,
  category: RewardCategory,
): Money | null {
  const card = sampleCards.find((c) => c.id === cardID)
  if (!card) return null
  const evaluation = evaluate(card, amount, '__base__', category, {
    merchants: sampleMerchants,
    offers: [],
    diningProgram: sampleDiningProgram,
  })
  return evaluation.totalValue
}
